Minify arbitrary values when printing candidates

RobinMalfait opened the pull request:
This PR will optimize and simplify the candidates when printing the candidate again after running codemods.

When we parse a candidate, we will add spaces around operators, for example `p-[calc(1px+1px)]]` will internally be handled as `calc(1px + 1px)`. Before this change, we would re-print this as: `p-[calc(1px_+_1px)]`.

This PR changes that by simplifying the candidate again so that the output is `p-[calc(1px+1px)]`. In addition, if _you_ wrote `p-[calc(1px_+_1px)]` then we will also simplify it to the concise form `p-[calc(1px_+_1px)]`.


Some examples:

Input:
```html
<div class="[p]:flex"></div>
<div class="[&:is(p)]:flex"></div>
<div class="has-[p]:flex"></div>
<div class="px-[theme(spacing.4)-1px]"></div>
```

Output before:
```html
<div class="[&:is(p)]:flex"></div>
<div class="[&:is(p)]:flex"></div>
<div class="has-[&:is(p)]:flex"></div>
<div class="px-[var(--spacing-4)_-_1px]"></div>
```

Output after:
```html
<div class="[p]:flex"></div>
<div class="[p]:flex"></div>
<div class="has-[p]:flex"></div>
<div class="px-[var(--spacing-4)-1px]"></div>
```

---

This is alternative implementation to #14717 and #14718
Closes: #14717 
Closes: #14718

RobinMalfait committed:
simplify candidate printing

RobinMalfait commented:
* **#14720** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14720?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `next`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @RobinMalfait and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

RobinMalfait commented on packages/@tailwindcss-upgrade/src/template/candidates.test.ts:
```diff
@@ -108,40 +108,63 @@ const candidates = [
-  'hover:peer-[&_p]:',
-  'hover:peer-[&_p]:focus:',
-  'peer-[&:hover]:peer-[&_p]:',
+  ['', ''], // no variant
```
Created a tuple with `in` and `out` (similar to the candidate above)

RobinMalfait committed:
update changelog

adamwathan commented on packages/@tailwindcss-upgrade/src/template/candidates.test.ts:
```diff
@@ -108,40 +108,63 @@ const candidates = [
+  ['min-[calc(1000px_+_12em)]:', 'min-[calc(1000px+12em)]:'],
+  ['min-[calc(1000px_+12em)]:', 'min-[calc(1000px+12em)]:'],
+  ['min-[calc(1000px+_12em)]:', 'min-[calc(1000px+12em)]:'],
+  ['min-[calc(1000px___+___12em)]:', 'min-[calc(1000px+12em)]:'],
```
It looks like we're missing tests for the cases where it is unsafe to remove the spaces (I don't know what these are off the top of my head but I know they exist).

Can we figure out what those cases are and make sure we handle them properly? Or do they actually not exist and CSS is just stupid for not letting us omit these stupid spaces? 😄

adamwathan replied:
Tried to look into this and I can't actually find any situations where it's ambiguous in reality? There were some talks about eventually supporting stuff like `calc(min-content * 50%)` and then the `-` in `min-content` is ambiguous as to whether it's subtraction or part of a keyword, but that isn't actually supported in CSS and it looks like `calc-size(…)` is what they are going to use for that stuff instead.

So fuck it, probably fine 😁

RobinMalfait replied:
I was thinking about this as well and very hard to come up with some tests.

I did see that we have `to-zero` as an edge case that we handle. Let me add a test for that.

adamwathan commented on packages/@tailwindcss-upgrade/src/template/candidates.ts:
```diff
@@ -138,3 +139,73 @@ function escapeArbitrary(input: string) {
     .replaceAll(' ', '_') // Replace spaces with underscores
 }
+
+function trimWhitespace(input: string) {
```
Looks like this and `escapeArbitrary` are never used separately, maybe should just make this one function like `toCandidateArbitraryValue` or something? Not totally sure on the name but really we're taking a CSS value and converting it to the arbitrary value of an actual class string.

Either way I don't think `trimWhitespace` is the right name since to me at least "trim" implies whitespace at the start and end, not whitespace inside the string.

RobinMalfait replied:
`minifyArbitraryValue` or `printArbitraryValue`. It's also used in variant and modifier positions so don't think the `candidate` in the name is a good idea.

Merging the functionality however is a good idea.

RobinMalfait replied:
Or `optimizeArbitraryValue`

RobinMalfait replied:
You know what, we already have `printVariant` and `printCandidate`, let's go with `printArbitraryValue`

RobinMalfait committed:
add additional test with `to-zero`

This is the only thing where I know we _don't_ want spaces. But this was
already handled when adding spaces internally.

RobinMalfait committed:
merge `escapeArbitrary` and `trimWhitespace` into `printArbitraryValue`

RobinMalfait committed:
move escape to the end

adamwathan commented on packages/@tailwindcss-upgrade/src/template/candidates.ts:
```diff
@@ -133,8 +134,75 @@ function printVariant(variant: Variant) {
+      if (idx === -1) return
+
+      let previous = parentArray[idx - 1]
+      if (previous?.kind !== 'separator' || previous.value !== ' ') return
```
What actually happens when we return `undefined` here? We're still interpolating the return value of this function into the class name at the call site.

RobinMalfait replied:
The surrounding `walk()` will keep walking (equivalent to `WalkAction.Continue`) because it could be that you have multiple `calc(…)` calls for example and if we can't upgrade the first one, we might be able to upgrade the second one.

RobinMalfait committed:
add failing test

RobinMalfait committed:
fix typo + add missing check

RobinMalfait committed:
ensure we handle ` / ` correctly
